megalinter/llm_provider/README.md

```diff
@@ -0,0 +1,143 @@
+| Google       | `llm_provider_google.py`      | Gemini Pro, Gemini Flash        | Yes          |
+| Mistral AI   | `llm_provider_mistral.py`     | Mistral Small, Medium, Large    | Yes          |
+| DeepSeek     | `llm_provider_deepseek.py`    | DeepSeek Chat, DeepSeek Coder   | Yes          |
+| Grok         | `llm_provider_grok.py`        | Grok Beta (xAI)                 | Yes          |
+| Hugging Face | `llm_provider_huggingface.py` | Any HF transformer model        | Optional     |
+| Ollama       | `llm_provider_ollama.py`      | Llama, CodeLlama, Mistral, etc. | No           |
+
+## Adding a New Provider
+
+1. **Create Provider Class**:
+   ```python
+   # llm_provider_newprovider.py
+   from .llm_provider import LLMProvider
+   
+   class NewProvider(LLMProvider):
+       def get_default_model(self) -> str:
+           return "default-model-name"
+       
+       def initialize(self) -> bool:
+           # Initialize your LLM here
+           # Return True if successful, False otherwise
+   ```
+
+2. **Register in Factory**:
+   ```python
+   # Add to llm_provider_factory.py
+   SUPPORTED_PROVIDERS = {
+       "newprovider": NewProvider,
+       # ... existing providers
+   }
+   ```
+
+3. **Update Imports**:
+   ```python
+   # Add to __init__.py
+   from .llm_provider_newprovider import NewProvider
+   ```
+
+## Provider Interface
+
+Each provider must implement:
+
+- `get_default_model() -> str`: Return the default model name
+- `initialize() -> bool`: Initialize the provider (return success status)
+- `load_config(request_id) -> Dict[str, Any]`: Load provider-specific configuration
+
+Base class provides:
+
+- `is_available() -> bool`: Check if provider is ready
+- `invoke(prompt, system_prompt=None) -> str`: Generate response
+- `get_config_value(key, default=None)`: Access configuration with centralized defaults
+- `get_default_config_value(key)`: Access centralized default values
+- `requires_api_key() -> bool`: Check if API key needed
+
+## Centralized Default Configuration
+
+The base class defines common default values that apply to all providers:
+
+```python
+DEFAULT_CONFIG = {
+    "temperature": 0.1,
+    "max_tokens": 1000,
+    "timeout": 30,
+    "retry_attempts": 3
+}
+```
+
+Providers automatically inherit these defaults through the `get_config_value()` method, ensuring consistency across all implementations.
+
+## Configuration
+
+Providers receive configuration through the constructor and use centralized defaults:
+
+```python
+config = {
+    "model_name": "model-name",
+    "temperature": 0.1,  # Uses centralized default if not specified
+    "max_tokens": 1000,  # Uses centralized default if not specified
+    "api_key": "your-api-key",
+    # Provider-specific settings...
+}
+provider = ProviderClass(config)
+```
+
+To use centralized defaults in your `load_config()` method:
+
+```python
+def load_config(self, request_id: str = None) -> Dict[str, Any]:
```

Comment:
Return type can be a bit more precise, as it mostly comes from env vars

```suggestion
def load_config(self, request_id: str = None) -> Dict[str, str | float]:
```

Note that I didn't add int too, as it is already covered by float, a ruff rule would trigger (if we have enough enabled and the one I think of is enabled, probably in the ANN category)


Also make sure that either a recent enough python is used or ```
from __future__ import annotations
```
To support the pipe character, also checked by a ruff rule (again, if we have enough enabled and the one I think of is enabled)